File: apps/android/domain/src/test/kotlin/care/pocket/domain/finance/FinanceVectors.kt
package care.pocket.domain.finance

import care.pocket.domain.vectors.FunctionRegistry
import care.pocket.domain.vectors.jsonNumber
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.double
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.long

// P1.3a: wires the real Finance.kt port into FunctionRegistry so
// finance.json's vectors un-skip. Registered under (domain="finance",
// fn=<name>) to match tools/golden-vectors/vectors/finance.json exactly.
// Field names and the money-string-vs-plain-number split below were read
// directly off tools/golden-vectors/export.ts's finance section, not
// guessed -- in particular, subscriptionImpact's and projectCashflow's
// money-shaped fields (totalPaid/opportunityCost/income/payments/...) are
// NOT stringified in the exported vectors (the exporter's rec() call sites
// for those two functions never wrap the result in amt()/mny()), unlike
// every other money-shaped result in this domain, which IS stringified.

private const val DOMAIN = "finance"

private fun JsonElement.asRecurringLike(): RecurringLike {
    val o = jsonObject
    return RecurringLike(o.getValue("amount").jsonPrimitive.long, o.getValue("frequency").jsonPrimitive.content)
}

private fun JsonElement.asCashflowInputs(): CashflowInputs {
    val o = jsonObject
    return CashflowInputs(
        monthlyIncome = o.getValue("monthlyIncome").jsonPrimitive.long,
        monthlyPayments = o.getValue("monthlyPayments").jsonPrimitive.long,
        monthlySavings = o.getValue("monthlySavings").jsonPrimitive.long,
        currentSavings = o.getValue("currentSavings").jsonPrimitive.long,
        annualReturnPct = o.getValue("annualReturnPct").jsonPrimitive.double,
        annualInflationPct = o.getValue("annualInflationPct").jsonPrimitive.double,
        incomeGrowthPct = o["incomeGrowthPct"]?.jsonPrimitive?.doubleOrNull ?: 0.0,
    )
}

private fun YearProjection.toJson(): JsonElement = JsonObject(
    mapOf(
        "year" to JsonPrimitive(year),
        "income" to JsonPrimitive(income),
        "payments" to JsonPrimitive(payments),
        "savingsContributed" to JsonPrimitive(savingsContributed),
        "netCashflow" to JsonPrimitive(netCashflow),
        "savingsBalance" to JsonPrimitive(savingsBalance),
        "realSavingsBalance" to JsonPrimitive(realSavingsBalance),
    )
)

private fun AmortRow.toJson(): JsonElement = JsonObject(
    mapOf(
        "month" to JsonPrimitive(month),
        "emi" to JsonPrimitive(emi.toString()),
        "interest" to JsonPrimitive(interest.toString()),
        "principal" to JsonPrimitive(principal.toString()),
        "balance" to JsonPrimitive(balance.toString()),
    )
)

private fun nullableString(e: JsonElement): String? =
    if (e is JsonNull) null else e.jsonPrimitive.contentOrNull

private fun nullableInt(e: JsonElement?): Int? =
    if (e == null || e is JsonNull) null else e.jsonPrimitive.int

fun registerFinanceVectors() {
    FunctionRegistry.register(DOMAIN, "futureValue") { input ->
        val o = input.jsonObject
        JsonPrimitive(
            futureValue(
                o.getValue("principal").jsonPrimitive.long,
                o.getValue("contribution").jsonPrimitive.long,
                o.getValue("periodicRate").jsonPrimitive.double,
                o.getValue("periods").jsonPrimitive.int,
            ).toString()
        )
    }

    FunctionRegistry.register(DOMAIN, "periodicRateFromAnnual") { input ->
        val o = input.jsonObject
        jsonNumber(periodicRateFromAnnual(o.getValue("annualPct").jsonPrimitive.double, o.getValue("period").jsonPrimitive.content))
    }

    FunctionRegistry.register(DOMAIN, "periodsToGoal") { input ->
        val o = input.jsonObject
        jsonNumber(
            periodsToGoal(
                o.getValue("current").jsonPrimitive.long,
                o.getValue("target").jsonPrimitive.long,
                o.getValue("contribution").jsonPrimitive.long,
                o.getValue("periodicRate").jsonPrimitive.double,
            )
        )
    }

    FunctionRegistry.register(DOMAIN, "monthlyEquivalent") { input ->
        val o = input.jsonObject
        JsonPrimitive(monthlyEquivalent(o.getValue("amount").jsonPrimitive.long, o.getValue("period").jsonPrimitive.content).toString())
    }

    FunctionRegistry.register(DOMAIN, "recurringMonthlyTotal") { input ->
        val o = input.jsonObject
        val items = o.getValue("items").jsonArray.map { it.asRecurringLike() }
        JsonPrimitive(recurringMonthlyTotal(items).toString())
    }

    FunctionRegistry.register(DOMAIN, "percentOfIncome") { input ->
        val o = input.jsonObject
        jsonNumber(percentOfIncome(o.getValue("monthlyAmount").jsonPrimitive.long, o.getValue("monthlyIncome").jsonPrimitive.long))
    }

    FunctionRegistry.register(DOMAIN, "subscriptionImpact") { input ->
        val o = input.jsonObject
        val r = subscriptionImpact(
            o.getValue("amount").jsonPrimitive.long,
            o.getValue("frequency").jsonPrimitive.content,
            o.getValue("years").jsonPrimitive.double,
            o.getValue("annualReturnPct").jsonPrimitive.double,
        )
        // NOT stringified -- see file header comment.
        JsonObject(mapOf("totalPaid" to JsonPrimitive(r.totalPaid), "opportunityCost" to JsonPrimitive(r.opportunityCost)))
    }

    FunctionRegistry.register(DOMAIN, "projectCashflow") { input ->
        val o = input.jsonObject
        val rows = projectCashflow(o.getValue("inp").asCashflowInputs(), o.getValue("years").jsonPrimitive.int)
        JsonArray(rows.map { it.toJson() })
    }

    FunctionRegistry.register(DOMAIN, "yearlyEquivalent") { input ->
        val o = input.jsonObject
        JsonPrimitive(yearlyEquivalent(o.getValue("amount").jsonPrimitive.long, o.getValue("period").jsonPrimitive.content).toString())
    }

    FunctionRegistry.register(DOMAIN, "emiFromPrincipal") { input ->
        val o = input.jsonObject
        JsonPrimitive(
            emiFromPrincipal(
                o.getValue("principal").jsonPrimitive.long,
                o.getValue("annualRatePct").jsonPrimitive.double,
                o.getValue("tenureMonths").jsonPrimitive.int,
            ).toString()
        )
    }

    FunctionRegistry.register(DOMAIN, "amortizationSchedule") { input ->
        val o = input.jsonObject
        val rows = amortizationSchedule(
            o.getValue("principal").jsonPrimitive.long,
            o.getValue("annualRatePct").jsonPrimitive.double,
            o.getValue("emi").jsonPrimitive.long,
            o.getValue("maxMonths").jsonPrimitive.int,
        )
        JsonArray(rows.map { it.toJson() })
    }

    FunctionRegistry.register(DOMAIN, "timeframeTotal") { input ->
        val o = input.jsonObject
        JsonPrimitive(timeframeTotal(o.getValue("monthlyAmount").jsonPrimitive.long, o.getValue("timeframe").jsonPrimitive.content).toString())
    }

    FunctionRegistry.register(DOMAIN, "emiDueDate") { input ->
        val o = input.jsonObject
        val result = emiDueDate(
            nullableString(o.getValue("startIso")),
            nullableInt(o["dueDay"]),
            o.getValue("emiNo").jsonPrimitive.int,
        )
        if (result == null) JsonNull else JsonPrimitive(result)
    }

    FunctionRegistry.register(DOMAIN, "isDuePassed") { input ->
        val o = input.jsonObject
        JsonPrimitive(isDuePassed(nullableString(o.getValue("dueIso")), o.getValue("asOfIso").jsonPrimitive.content))
    }

    FunctionRegistry.register(DOMAIN, "effectivePaidEmis") { input ->
        val o = input.jsonObject
        val manual = o.getValue("manual").jsonArray.map { it.jsonPrimitive.int }
        val opts = o["opts"]?.jsonObject
        val result = effectivePaidEmis(
            manual = manual,
            totalEmis = o.getValue("totalEmis").jsonPrimitive.int,
            autoMark = opts?.get("autoMark")?.jsonPrimitive?.boolean ?: false,
            startIso = opts?.get("startIso")?.let { nullableString(it) },
            dueDay = opts?.get("dueDay")?.let { nullableInt(it) },
            asOfIso = opts?.get("asOfIso")?.jsonPrimitive?.content ?: "1970-01-01",
        )
        JsonArray(result.sorted().map { JsonPrimitive(it) })
    }
}

File: apps/ios/Domain/Tests/DomainTests/FinanceVectors.swift
import Foundation
@testable import Domain

// P1.3b: wires the real Finance.swift port into FunctionRegistry so
// finance.json's vectors un-skip. Registered under (domain="finance",
// fn=<name>) to match tools/golden-vectors/vectors/finance.json exactly --
// mirrors the Android adapter (FinanceVectors.kt) field-for-field,
// including its money-string-vs-plain-number split (subscriptionImpact's
// and projectCashflow's money-shaped fields are NOT stringified in the
// exported vectors, unlike every other money-shaped result in this
// domain -- verified against tools/golden-vectors/export.ts's finance
// section, not guessed).

/// Infinity -> NSNull (mirrors JS's JSON.stringify(Infinity) === "null");
/// everything else -> NSNumber. Needed for periodicRateFromAnnual,
/// periodsToGoal, percentOfIncome (this domain is the first native port
/// to actually produce Infinity results). Internal (not private), not
/// file-scoped -- BudgetVectors.swift's budgetProgress.pct reuses it
/// rather than duplicating the same Infinity-handling logic twice.
///
/// **It was called `jsonNumber` and had to be renamed.** Domain later gained a
/// PUBLIC `jsonNumber(Double) -> String`, and inside this test target the
/// target's own declaration wins over the imported module's. Every
/// `jsonNumber(...)` in AssistantVectors silently called THIS instead — no
/// ambiguity error, no warning, just 41 vectors comparing a number against a
/// string. A same-named helper in a test target shadowing a public API is a
/// silent-wrong-value trap, so the local one moved.
func jsonOrNull(_ n: Double) -> Any {
    n.isFinite ? NSNumber(value: n) : NSNull()
}

private func asRecurringLike(_ any: Any) -> RecurringLike {
    let d = any as! [String: Any]
    return RecurringLike(amount: (d["amount"] as! NSNumber).int64Value, frequency: d["frequency"] as! String)
}

private func asCashflowInputs(_ any: Any) -> CashflowInputs {
    let d = any as! [String: Any]
    return CashflowInputs(
        monthlyIncome: (d["monthlyIncome"] as! NSNumber).int64Value,
        monthlyPayments: (d["monthlyPayments"] as! NSNumber).int64Value,
        monthlySavings: (d["monthlySavings"] as! NSNumber).int64Value,
        currentSavings: (d["currentSavings"] as! NSNumber).int64Value,
        annualReturnPct: (d["annualReturnPct"] as! NSNumber).doubleValue,
        annualInflationPct: (d["annualInflationPct"] as! NSNumber).doubleValue,
        incomeGrowthPct: (d["incomeGrowthPct"] as? NSNumber)?.doubleValue ?? 0
    )
}

private func yearProjectionToJson(_ p: YearProjection) -> [String: Any] {
    [
        "year": p.year,
        "income": p.income,
        "payments": p.payments,
        "savingsContributed": p.savingsContributed,
        "netCashflow": p.netCashflow,
        "savingsBalance": p.savingsBalance,
        "realSavingsBalance": p.realSavingsBalance,
    ]
}

private func amortRowToJson(_ r: AmortRow) -> [String: Any] {
    [
        "month": r.month,
        "emi": String(r.emi),
        "interest": String(r.interest),
        "principal": String(r.principal),
        "balance": String(r.balance),
    ]
}

private func nullableString(_ any: Any?) -> String? {
    guard let any, !(any is NSNull) else { return nil }
    return any as? String
}

private func nullableInt(_ any: Any?) -> Int? {
    guard let any, !(any is NSNull) else { return nil }
    return (any as? NSNumber)?.intValue
}

func registerFinanceVectors() {
    FunctionRegistry.register(domain: "finance", fn: "futureValue") { input in
        let d = input as! [String: Any]
        let v = try futureValue(
            (d["principal"] as! NSNumber).int64Value,
            (d["contribution"] as! NSNumber).int64Value,
            (d["periodicRate"] as! NSNumber).doubleValue,
            (d["periods"] as! NSNumber).intValue
        )
        return String(v)
    }

    FunctionRegistry.register(domain: "finance", fn: "periodicRateFromAnnual") { input in
        let d = input as! [String: Any]
        return jsonOrNull(periodicRateFromAnnual((d["annualPct"] as! NSNumber).doubleValue, d["period"] as! String))
    }

    FunctionRegistry.register(domain: "finance", fn: "periodsToGoal") { input in
        let d = input as! [String: Any]
        return jsonOrNull(
            periodsToGoal(
                (d["current"] as! NSNumber).int64Value,
                (d["target"] as! NSNumber).int64Value,
                (d["contribution"] as! NSNumber).int64Value,
                (d["periodicRate"] as! NSNumber).doubleValue
            )
        )
    }

    FunctionRegistry.register(domain: "finance", fn: "monthlyEquivalent") { input in
        let d = input as! [String: Any]
        return String(monthlyEquivalent((d["amount"] as! NSNumber).int64Value, d["period"] as! String))
    }

    FunctionRegistry.register(domain: "finance", fn: "recurringMonthlyTotal") { input in
        let d = input as! [String: Any]
        let items = (d["items"] as! [Any]).map(asRecurringLike)
        return String(recurringMonthlyTotal(items))
    }

    FunctionRegistry.register(domain: "finance", fn: "percentOfIncome") { input in
        let d = input as! [String: Any]
        return jsonOrNull(percentOfIncome((d["monthlyAmount"] as! NSNumber).int64Value, (d["monthlyIncome"] as! NSNumber).int64Value))
    }

    FunctionRegistry.register(domain: "finance", fn: "subscriptionImpact") { input in
        let d = input as! [String: Any]
        let r = try subscriptionImpact(
            (d["amount"] as! NSNumber).int64Value,
            d["frequency"] as! String,
            (d["years"] as! NSNumber).doubleValue,
            (d["annualReturnPct"] as! NSNumber).doubleValue
        )
        // NOT stringified -- see file header comment.
        return ["totalPaid": r.totalPaid, "opportunityCost": r.opportunityCost]
    }

    FunctionRegistry.register(domain: "finance", fn: "projectCashflow") { input in
        let d = input as! [String: Any]
        let rows = try projectCashflow(asCashflowInputs(d["inp"]!), (d["years"] as! NSNumber).intValue)
        return rows.map(yearProjectionToJson)
    }

    FunctionRegistry.register(domain: "finance", fn: "yearlyEquivalent") { input in
        let d = input as! [String: Any]
        return String(yearlyEquivalent((d["amount"] as! NSNumber).int64Value, d["period"] as! String))
    }

    FunctionRegistry.register(domain: "finance", fn: "emiFromPrincipal") { input in
        let d = input as! [String: Any]
        return String(
            emiFromPrincipal(
                (d["principal"] as! NSNumber).int64Value,
                (d["annualRatePct"] as! NSNumber).doubleValue,
                (d["tenureMonths"] as! NSNumber).intValue
            )
        )
    }

    FunctionRegistry.register(domain: "finance", fn: "amortizationSchedule") { input in
        let d = input as! [String: Any]
        let rows = amortizationSchedule(
            (d["principal"] as! NSNumber).int64Value,
            (d["annualRatePct"] as! NSNumber).doubleValue,
            (d["emi"] as! NSNumber).int64Value,
            (d["maxMonths"] as! NSNumber).intValue
        )
        return rows.map(amortRowToJson)
    }

    FunctionRegistry.register(domain: "finance", fn: "timeframeTotal") { input in
        let d = input as! [String: Any]
        return String(timeframeTotal((d["monthlyAmount"] as! NSNumber).int64Value, d["timeframe"] as! String))
    }

    FunctionRegistry.register(domain: "finance", fn: "emiDueDate") { input in
        let d = input as! [String: Any]
        let result = emiDueDate(nullableString(d["startIso"]), nullableInt(d["dueDay"]), (d["emiNo"] as! NSNumber).intValue)
        if let result { return result }
        return NSNull()
    }

    FunctionRegistry.register(domain: "finance", fn: "isDuePassed") { input in
        let d = input as! [String: Any]
        return isDuePassed(nullableString(d["dueIso"]), d["asOfIso"] as! String)
    }

    FunctionRegistry.register(domain: "finance", fn: "effectivePaidEmis") { input in
        let d = input as! [String: Any]
        let manual = (d["manual"] as! [Any]).map { ($0 as! NSNumber).intValue }
        let opts = d["opts"] as? [String: Any]
        let result = effectivePaidEmis(
            manual: manual,
            totalEmis: (d["totalEmis"] as! NSNumber).intValue,
            autoMark: (opts?["autoMark"] as? NSNumber)?.boolValue ?? false,
            startIso: nullableString(opts?["startIso"]),
            dueDay: nullableInt(opts?["dueDay"]),
            asOfIso: (opts?["asOfIso"] as? String) ?? "1970-01-01"
        )
        return result.sorted()
    }

    FunctionRegistry.register(domain: "finance", fn: "chargesToDate") { input in
        let d = input as! [String: Any]
        return chargesToDate(nullableString(d["startIso"]), d["period"] as! String, d["asOfIso"] as! String)
    }

    FunctionRegistry.register(domain: "finance", fn: "estimatedSpentToDate") { input in
        let d = input as! [String: Any]
        return String(estimatedSpentToDate(
            (d["amount"] as! NSNumber).int64Value,
            nullableString(d["startIso"]),
            d["period"] as! String,
            d["asOfIso"] as! String
        ))
    }
}
